File: read_hdf5.m
function [Time, output] = read_hdf5(device,shot_number,channel )
% author:huangyao
% email: hyao666@foxmail.com
dir_name = fix(shot_number / 200) * 200;
hl2a = {'HL-2A', 'HL2A', '2A'};
jtext = {'J-TEXT', 'JTEXT'};
east = {'EAST'};
device = upper(device);

Time=0;
output=0;
% ¼ÓÔØhdf5ÎÄ¼þ
hinfo = false;
try    
    if strcmpi(device,hl2a(1)) || strcmpi(device, hl2a(2)) || strcmpi(device,hl2a(3))
        hinfo = hdf5info(['\\192.168.9.242\hdf\2A/',num2str(dir_name),'/HL-2A=',num2str(shot_number),'=PhysicsDB.H5']);
    elseif strcmpi(device,jtext(1)) || strcmpi(device, jtext(2)) 
        hinfo = hdf5info(['\\192.168.9.242\hdf\J-TEXT/',num2str(dir_name),'/JTEXT=',num2str(shot_number),'=PhysicsDB.h5']);
    elseif strcmpi(device,east)
        hinfo = hdf5info(['\\192.168.9.242\hdf\EAST/',num2str(dir_name),'/EAST=',num2str(shot_number),'=PhysicsDB.h5']);
    end
    
catch 
    disp('No such file or directory')
    return 
end

if ~hinfo
    disp('please input right device name:')
    disp('such as:2a,jtext,east....')
    return
end
    

Data=hinfo.GroupHierarchy.Groups;

end1 = size(Data);
break_mark=false;
for index1 = 1:end1(2)
    Groups=Data(index1).Datasets;
    end2 = size(Groups);
    
    for index2 =1:end2(2)
%         Datasets=Groups(index2);
        name= Groups(index2).Name;
        
        if strfind(name,channel)
            break_mark = true;
            break
        end
        
    end
    
    if break_mark
        break
    end
end

if ~break_mark
    disp('can not find channel')
    return
end
clear end1 end2 break_mark name Groups Datasets

Data=hinfo.GroupHierarchy.Groups(index1).Datasets(index2);
output=hdf5read(Data);

data_size=size(Data.Attributes);
for index=1:data_size(2)
    name = Data.Attributes(index).Name;
    if strfind(name,'T_Start')
        T_Start = Data.Attributes(index).Value;
    elseif strfind(name,'T_Freq')
        T_Freq = Data.Attributes(index).Value;
    end
end


data_point_num = size(output);
data_point_num=data_point_num(1);
Time = (T_Start:data_point_num-1) / double(T_Freq);
Time=Time';
return

File: read_hdf5.py
# author:yellowyellowyao
from __future__ import print_function
import h5py
import numpy as np


def read_hdf5(device,shot_number,channel):

    dir_name = shot_number // 200 * 200

    hl2a = ["HL-2A", "HL2A", "2A"]
    jtext = ["J-TEXT", "JTEXT"]
    east = ["EAST"]
    device = device.upper()

    # 加载hdf5文件
    try:
        if device in hl2a:
            Data = h5py.File(r"\\192.168.9.242\hdf\2A/"+str(dir_name)+"/HL-2A="+str(shot_number)+"=PhysicsDB.H5","r")
        elif device in jtext :
            Data = h5py.File(r"\\192.168.9.242\hdf\J-TEXT/"+str(dir_name)+"/JTEXT="+str(shot_number)+"=PhysicsDB.h5","r")
        elif device in east:
            Data = h5py.File(r"\\192.168.9.242\hdf\EAST/"+str(dir_name)+"/EAST="+str(shot_number)+"=PhysicsDB.h5","r")
    except OSError:
            print('No such file or directory')
            return 0, 0

    # 读取信号
    output= []
    for para in Data:
        try:
            output = Data[para][channel]
            break
        except:
            continue

    if not output:
        print("can't find channel")
        return 0,0

    output = np.array(output)

    # 读取信号的附加信息
    T_Start = Data[para][channel].attrs['T_Start']
    T_Freq = Data[para][channel].attrs['T_Freq']
    data_point_num = output.shape[0]
    Time = np.arange(T_Start, data_point_num) / T_Freq

    return Time, output


#################### example ####################
# import matplotlib.pyplot as plt
# plt.rcParams['font.sans-serif'] = ['SimHei']  # 用来正常显示中文标签
# plt.rcParams['axes.unicode_minus'] = False    # 用来正常显示负号

# Time, Channel = read_hdf5("2a",10022,"IP")
# plt.gcf().set_size_inches(16, 8)
# plt.plot(Time, Channel)
# plt.grid(True)
# plt.show()
